File: lib/app/logging/privacy_safe_log.dart
import 'package:flutter/foundation.dart';

abstract final class PrivacySafeLog {
  static final RegExp _safeEvent = RegExp(r'^[a-z0-9_]{1,80}$');
  static final RegExp _safeToken = RegExp(r'^[A-Za-z0-9_.-]{1,80}$');
  static const Set<String> _allowedFields = <String>{
    'category',
    'count',
    'status',
    'examined',
    'deleted',
    'failed',
    'hasMore',
  };

  static void info(String event, [Map<String, Object?> fields = const {}]) {
    debugPrint(format(event, fields));
  }

  @visibleForTesting
  static String format(String event, [Map<String, Object?> fields = const {}]) {
    final safeName = _safeEvent.hasMatch(event) ? event : 'invalid_event';
    final entries =
        fields.entries
            .where(
              (entry) =>
                  _allowedFields.contains(entry.key) && entry.value != null,
            )
            .toList(growable: false)
          ..sort((left, right) => left.key.compareTo(right.key));
    final encoded = entries
        .map((entry) => '${entry.key}=${_safeValue(entry.value!)}')
        .join(' ');
    return encoded.isEmpty ? 'event=$safeName' : 'event=$safeName $encoded';
  }

  static String _safeValue(Object value) {
    if (value is num || value is bool) return value.toString();
    final text = value.toString();
    return _safeToken.hasMatch(text) ? text : 'redacted';
  }
}

File: apps/mobile-flutter/android/app/src/main/kotlin/com/voice2text/app/privacy/PrivacySafeLog.kt
package com.voice2text.app.privacy

import android.util.Log

object PrivacySafeLog {
    private val safeToken = Regex("[A-Za-z0-9_.-]{1,80}")
    private val safeEvent = Regex("[a-z0-9_]{1,80}")
    private val allowedFields =
        setOf(
            "category",
            "count",
            "jobId",
            "model",
            "durationMs",
            "costMs",
            "stage",
        )

    fun info(
        tag: String,
        event: String,
        fields: Map<String, Any?> = emptyMap(),
    ) {
        Log.i(tag, format(event, fields))
    }

    fun error(
        tag: String,
        event: String,
        fields: Map<String, Any?> = emptyMap(),
    ) {
        Log.e(tag, format(event, fields))
    }

    fun format(
        event: String,
        fields: Map<String, Any?> = emptyMap(),
    ): String {
        val safeName = event.takeIf(safeEvent::matches) ?: "invalid_event"
        val encoded =
            fields.entries
                .asSequence()
                .filter { it.key in allowedFields && it.value != null }
                .sortedBy { it.key }
                .joinToString(" ") { (key, value) ->
                    "$key=${safeValue(value ?: "redacted")}"
                }
        return if (encoded.isEmpty()) "event=$safeName" else "event=$safeName $encoded"
    }

    private fun safeValue(value: Any): String =
        when (value) {
            is Number, is Boolean -> value.toString()
            else -> value.toString().takeIf(safeToken::matches) ?: "redacted"
        }
}
